File: src/golang/9663.go
// 백준 9663번 문제 : N-Queen
// https://www.acmicpc.net/problem/9663
// 알고리즘 분류 : 브루트포스 알고리즘, 백트래킹

package main

import (
	"bufio"
	"fmt"
	"os"
)

var n int
var result int = 0
var diagonalSize int

func main() {
	var reader *bufio.Reader = bufio.NewReader(os.Stdin)
	var writer *bufio.Writer = bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	fmt.Fscan(reader, &n)

	diagonalSize = n*2 - 1
	colStatus := make([]bool, n)
	diagonalStatus := make([]bool, diagonalSize)
	skewDiagonalStatus := make([]bool, diagonalSize)

	nQueen(0, &colStatus, &diagonalStatus, &skewDiagonalStatus)

	fmt.Fprintln(writer, result)
}

func nQueen(row int, colStatus *[]bool,
	diagonalStatus *[]bool, skewDiagonalStatus *[]bool) {
	if row == n {
		return
	}

	for i := 0; i < n; i++ {
		if (*colStatus)[i] || (*diagonalStatus)[n-1+row-i] || (*skewDiagonalStatus)[row+i] {
			continue
		}

		(*colStatus)[i] = true
		(*diagonalStatus)[n-1+row-i] = true
		(*skewDiagonalStatus)[row+i] = true

		if row == n-1 {
			result += 1
		} else {
			nQueen(row+1, colStatus, diagonalStatus, skewDiagonalStatus)
		}

		(*skewDiagonalStatus)[row+i] = false
		(*diagonalStatus)[n-1+row-i] = false
		(*colStatus)[i] = false
	}
}

File: src/c/9663.c
// 백준 9663번 문제 : N-Queen
// https://www.acmicpc.net/problem/9663
// 알고리즘 분류 : 브루트포스 알고리즘, 백트래킹

#include <stdio.h>
#include <stdbool.h>

int n = 0;
bool is_used_column[40];
bool is_used_right_up[40];
bool is_used_right_down[40];
int result = 0;

void n_queen(int cur)
{
    if (cur == n)
    {
        result++;
        return;
    }

    for (int i = 0; i < n; i++)
    {
        // Yes. It is an impossible case
        if (is_used_column[i] || is_used_right_up[i + cur] ||
            is_used_right_down[n - 1 + i - cur]) continue;
        is_used_column[i] = true;
        is_used_right_up[i + cur] = true;
        is_used_right_down[n - 1 + i - cur] = true;

        // Go to the next row
        n_queen(cur + 1);

        // Flag state to false
        is_used_right_down[n - 1 + i - cur] = false;
        is_used_right_up[i + cur] = false;
        is_used_column[i] = false;
    }
}

int main(void)
{
    scanf("%d", &n);

    n_queen(0);
    printf("%d\n", result);

    return 0;
}
